src/tuple_utils.c

```diff
@@ -0,0 +1,21 @@
+/*
+ * Copyright (c) 2016-2018  Timescale, Inc. All Rights Reserved.
+ *
+ * This file is licensed under the Apache License,
+ * see LICENSE-APACHE at the top level directory.
+ */
+#include <postgres.h>
+#include <access/htup_details.h>
+
+#include "tuple_utils.h"
+
+/* Create a fresh struct pointer that will contain copied contents of the tuple */
+void *
+create_struct_from_tuple(HeapTuple tuple, MemoryContext mctx, size_t alloc_size, size_t copy_size)
```

Comment:
I would actually suggest wrapping this as a macro to avoid type casting when calling this fucntion. Something like:

```
#define create_struct_from_tuple(tuple, to_type, form_type, mctx) \
      (to_type *) create_struct_from_tuple_internal(type, mctx, sizeof(to_type), sizeof(form_type));
```

Also note that `GETSTRUCT()` only works for tuples with fixed size members (or the initial fixed size ones). If there are varlena types, this function can't be used. I think that should at least be clarified in the comments.

Reply:
Created the macro and fixed comment. Although, it's not like we really define any FormData types that are varlena in our codebase?

Reply:
FWIW, I think this also applies to tuples that can have NULL values (i.e., if a value is NULL the tuple is compressed and `GETSTRUCT()` breaks.) This is the case for, e.g., the dimension table.